Add hub integration

davidberenstein1957 opened the pull request:
I added a basic integration for the Hub.

```python
import numpy as np

from vicinity import Backend, Metric, Vicinity

# Create some dummy data
items = [
    {"name": "triforce", "id": 0},
    {"name": "master sword", "id": 1},
    {"name": "hylian shield", "id": 2},
    {"name": "boomerang", "id": 3},
    {"name": "hookshot", "id": 4},
]
vectors = np.random.rand(len(items), 128)

# Initialize the Vicinity instance (using the basic backend and cosine metric)
vicinity = Vicinity.from_vectors_and_items(vectors=vectors, items=items)
vicinity.push_to_hub("davidberenstein1957/my-vicinity-repo")
vicinity = Vicinity.load_from_hub("davidberenstein1957/my-vicinity-repo")
```

Resulting in https://huggingface.co/datasets/davidberenstein1957/my-vicinity-repo
You can also find datasets here: https://huggingface.co/datasets?other=vicinity

Pringled commented on vicinity/integrations/huggingface.py:
```diff
@@ -0,0 +1,162 @@
+            )
+
+        # DatasetCard
+        DatasetCard(
```
Would move this to a file dataset_card_template.md, bit cleaner imo

Pringled commented on vicinity/integrations/huggingface.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        Save the Vicinity instance to the Hugging Face Hub.
+
+        Args:
```
Minor suggestion but the docstrings are inconsistent (`load_from_hub` uses the docstring format that we usually use).

Pringled commented on vicinity/integrations/huggingface.py:
```diff
@@ -0,0 +1,162 @@
+
+        :param repo_id: The repository ID on the Hugging Face Hub.
+        :param token: Optional authentication token for private repositories.
+        :param kwargs: Additional arguments passed to load_dataset.
```
```suggestion
        :param **kwargs: Additional arguments passed to load_dataset.
```
Otherwise pre-commit will complain

Pringled commented on vicinity/integrations/huggingface.py:
```diff
@@ -0,0 +1,162 @@
+import json
```
Add `from __future__ import annotations` so | is supported for typing (python 3.9 support)

Pringled commented on vicinity/integrations/huggingface.py:
```diff
@@ -0,0 +1,162 @@
+
+
+class HuggingFaceMixin:
+    def save_to_hub(
```
This just calls push_to_hub right? Is there a specific reason for not directly calling push_to_hub, or having the logic live in this function?

Pringled commented on vicinity/vicinity.py:
```diff
@@ -19,8 +20,13 @@
 
 logger = logging.getLogger(__name__)
 
+if importlib.util.find_spec("huggingface_hub") is not None and importlib.util.find_spec("datasets") is not None:
```
This should be an optional dependency group imo (can be `huggingface`, or `integrations`, or both and `integration` installs all integrations). Right now, I think if someone has not installed these libs and tries to run `vicinity.save_to_hub()`, it will just crash, it would be nicer if there's a specific error thrown to show that the optional dependency is not installed. Similar to this: https://github.com/MinishLab/model2vec/blob/main/model2vec/distill/__init__.py#L5

Pringled reviewed (changes requested):
Really nice integration, would be awesome to have this in Vicinity 🎉 Left some minor feedback

Pringled commented:
Also feel free to add an example to the main README (e.g. in the quickstart there's a saving/loading part, it can be added under there)

davidberenstein1957 commented:
- I resolved the comments.
- Added a basic test for testing `load_from_hub`. Did not add the `push_to_hub` integration because I did not want to fidget with tokens. Perhaps we can use the dataset we include for testing and transfer it to your Hub org?
- Added a `model_name_or_path` argument because vector indices aren't sharable/usable otherwise.

Pringled reviewed (approved):
LGTM, very nice! I will put the test dataset on our org as well and do a small followup to change the path. Merging!
